### ISSUE TITLE
Create rolebindings for Kubernetes deployment if RBAC enabled

### ISSUE THREAD

**Issue by calohmn** [EnMasseProject/enmasse#1271]
On clusters with RBAC support enabled, the Kubernetes deployment with `deploy.sh` will not work because of missing rolebindings.

As also minikube seems to have RBAC enabled by default now (tested with 0.27.0), it would be good to create the necessary rolebindings in `deploy.sh` if RBAC is enabled.

Related issue: #201.

**Comment by calohmn** [comment]
It would be good to use more fine-grained permissions than the ones in #201 / PR #203.

Adding a rolebinding to the admin clusterrole
````
kubectl create rolebinding enmasse-admin-binding --clusterrole=admin --serviceaccount=${NAMESPACE}:enmasse-admin --namespace=${NAMESPACE}
````
leads to one remaining issue in the address-space-controller:
````
2018-05-24 16:52:06 WARN  ControllerChain:124 - Error syncing address space default
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1/. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. forbidden: User "system:serviceaccount:enmasse:enmasse-admin" cannot get path "/".
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:470)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:407)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:379)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:343)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:327)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getRootPaths(BaseOperation.java:236)
	at io.fabric8.kubernetes.client.BaseClient.rootPaths(BaseClient.java:140)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isOpenShift(OpenshiftAdapterSupport.java:53)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isOpenShiftAPIGroups(OpenshiftAdapterSupport.java:92)
	at io.fabric8.openshift.client.OpenShiftConfig.isOpenShiftAPIGroups(OpenShiftConfig.java:120)
	at io.fabric8.openshift.client.dsl.internal.OpenShiftOperation.withApiGroup(OpenShiftOperation.java:55)
	at io.fabric8.openshift.client.dsl.internal.OpenShiftOperation.withApiGroup(OpenShiftOperation.java:78)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:82)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:73)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:69)
	at io.fabric8.openshift.client.DefaultOpenShiftClient.templates(DefaultOpenShiftClient.java:372)
	at io.enmasse.controller.common.KubernetesHelper.processTemplate(KubernetesHelper.java:124)
	at io.enmasse.controller.TemplateInfraResourceFactory.createResourceList(TemplateInfraResourceFactory.java:128)
	at io.enmasse.controller.CreateController.handle(CreateController.java:67)
	at io.enmasse.controller.ControllerChain.onUpdate(ControllerChain.java:119)
	at io.enmasse.k8s.api.ResourceChecker.doWork(ResourceChecker.java:48)
	at io.enmasse.k8s.api.ResourceChecker.run(ResourceChecker.java:40)
	at java.lang.Thread.run(Thread.java:748)
````

This was already mentioned in #201 and fabric8io/kubernetes-client#850.

I tried using an additional ClusterRole 
````
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: get-root-paths
rules:
- nonResourceURLs: ["/"]
  verbs: ["get", "list"]
````
with corresponding binding for that
````
kubectl create rolebinding enmasse-admin-get-root-paths-binding --clusterrole=get-root-paths --serviceaccount=${NAMESPACE}:enmasse-admin --namespace=${NAMESPACE}
````
, but for whatever reason this doesn't seem to work (minikube 0.27.0, k8s v1.9.4 and 1.10.0).

**Comment by calohmn** [comment]
@lulf looking at the newly added files in the "cluster-roles" directory, these are only geared towards Openshift, first and foremost concerning the apiVersion:
https://github.com/EnMasseProject/enmasse/blob/0d38f01e0aa8f46518148fd11f6d1ed85d8aa877/templates/include/cluster-roles/api-server.yaml#L1
Using `v1` leads to an error when applying it on K8s 1.9. On Kubernetes, `rbac.authorization.k8s.io/v1` looks like the apiVersion to be used.

Can we use `rbac.authorization.k8s.io/v1` for OpenShift here as well? 
The OpenShift docs aren't clear to me in that regard (see [Origin](https://docs.openshift.org/latest/rest_api/apis-rbac.authorization.k8s.io/v1.ClusterRoleBinding.html) vs. [3.9](https://docs.openshift.com/container-platform/3.9/rest_api/apis-rbac.authorization.k8s.io/v1beta1.ClusterRole.html)).

**Comment by lulf** [comment]
@calohmn Yes, Openshift 3.9 supports the same. However, we need to upgrade our CI from 3.7 first :)

**Comment by calohmn** [comment]
@lulf When testing a fix here on minikube with RBAC enabled, where I've added admin (cluster)rolebindings for the `enmasse-admin` and `address-space-admin` service accounts, there is this output in the k8s-api-server log:
````
I0613 16:01:09.942190       1 rbac.go:116] RBAC DENY: user "system:anonymous" groups ["system:unauthenticated"] cannot "get" resource "configmaps" named "keycloak-config" in namespace "hono"
I0613 16:01:09.946724       1 rbac.go:116] RBAC DENY: user "system:anonymous" groups ["system:unauthenticated"] cannot "list" resource "configmaps" in namespace "hono"
````
and this output in the `AddressSpaceController` log:
````
2018-06-13 16:01:10 WARN  Main:105 - Skipping standard authentication service: configmap keycloak-config not found
````

I wonder why the "system:anonymous" user and not the namespace-specific `default` service account or the `enmasse-admin` serviceaccount is used here from what seems like `io.enmasse.controller.Main.controllerClient`.

The last test I've done was with master, so including these changes:
https://github.com/EnMasseProject/enmasse/commit/10702200491d221648f7fdbe6ab2288ace9aa04c#diff-907685d9dd995c8890fbdff2604710d4R34

Also testing with 0.20.0, i.e. the code where `DefaultOpenShiftClient` is initialized with OauthToken and Namespace, there is the error. Maybe "withUsername" is needed here?

**Comment by lulf** [comment]
@calohmn I think there are more stuff that needs to be changed. Basically the address space controller KubernetesHelper contains a lot of 

```
if (client.isAdaptable(OpenShiftClient.class)) {
  // Openshift specific RBAC stuff
} else {
  // do nothing
}
```

I think we should fix that as part of moving to the 1.9/3.9 rbac resources

**Comment by calohmn** [comment]
@lulf Ok. So it looks like it's about these methods in `KubernetesHelper`:
https://github.com/EnMasseProject/enmasse/blob/981d1ce8dec7d5d6ffd4fddaf6642e106a2128b5/address-space-controller/src/main/java/io/enmasse/controller/CreateController.java#L66-L68

Regarding the `RBAC DENY` log entries above: These are a bit misleading, as they don't correspond to errors on the EnMasse level. It's just that the `DefaultOpenShiftClient` obviously first does an unauthenticated request and only then a retry with the Authorization header (see [addressSpaceController_excerpt.log](https://github.com/EnMasseProject/enmasse/files/2104942/addressSpaceController_excerpt.log)). Later requests are all using authorization then.

(Concerning the repeated log lines with `RBAC DENY: user "system:serviceaccount:hono:default" groups ["system:serviceaccounts" "system:serviceaccounts:hono" "system:authenticated"] cannot "list" resource "pods" in namespace "hono"`, I've created #1367.)

**Comment by calohmn** [comment]
@lulf After merge of your PR #1377, what remains to be done now is having a ClusterRole with
`"rules":[{"nonResourceURLs":["/"],"verbs":["get"]}]`
and creating ClusterRoleBindings with it for the `enmasse-admin` and `address-space-admin` service accounts (in `deploy.sh` and `KubernetesHelper.addAddressSpaceRoleBindings()`).
I was anyway just working on that and can create a PR.

**Comment by lulf** [comment]
@calohmn I don't think its needed in KubernetesHelper.addAddressSpaceRoleBindings()?

I'd also like to avoid adding it there, because I don't think we should rely on being able to create that kind of clusterrole in all environments.

For openshift this clusterrole is not needed, so preferrably it should only be created running on kubernetes.

**Comment by calohmn** [comment]
@lulf I can understand that this should be avoided if possible. On the other hand, the clusterrole is needed for the `address-space-admin` service-account, otherwise there is this exception (in the `standard-controller` when creating the first brokered address):
````
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. forbidden: User "system:serviceaccount:hono:address-space-admin" cannot get path "/".
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:470)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:407)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:379)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:343)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:327)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getRootPaths(BaseOperation.java:236)
	at io.fabric8.kubernetes.client.BaseClient.rootPaths(BaseClient.java:140)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isOpenShift(OpenshiftAdapterSupport.java:53)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isOpenShiftAPIGroups(OpenshiftAdapterSupport.java:92)
	at io.fabric8.openshift.client.OpenShiftConfig.isOpenShiftAPIGroups(OpenShiftConfig.java:120)
	at io.fabric8.openshift.client.dsl.internal.OpenShiftOperation.withApiGroup(OpenShiftOperation.java:55)
	at io.fabric8.openshift.client.dsl.internal.OpenShiftOperation.withApiGroup(OpenShiftOperation.java:78)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:82)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:73)
	at io.fabric8.openshift.client.dsl.internal.TemplateOperationsImpl.<init>(TemplateOperationsImpl.java:69)
	at io.fabric8.openshift.client.DefaultOpenShiftClient.templates(DefaultOpenShiftClient.java:372)
	at io.enmasse.controller.standard.KubernetesHelper.processTemplate(KubernetesHelper.java:141)
````
In multitenant-mode, the `address-space-admin` service-account is only created from KubernetesHelper. 
So that means the ClusterRoleBinding has to be created there as well.

Or should we have this operation in `standard-controller` be done with the `enmasse-admin` service-account?

**Comment by calohmn** [comment]
@lulf I think the 3.9 change probably doesn't make much of a difference here. I've created #1382 with changes for singletenant and multitenant modes.

**Comment by lulf** [comment]
Fixed in 0.23.0
